// @iden3/js-merkletree/dist/cjs/lib/merkletree/proof.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Proof = void 0;
const constants_1 = require("../../constants");
const utils_1 = require("../utils");
class Proof {
    constructor() {
        this.existence = false;
        this.depth = 0;
        this.siblings = [];
        const arrBuff = new ArrayBuffer(constants_1.NOT_EMPTIES_LEN);
        this.notEmpties = new Uint8Array(arrBuff);
    }
    bytes() {
        let bsLen = constants_1.PROOF_FLAG_LEN + this.notEmpties.length + constants_1.ELEM_BYTES_LEN * this.siblings.length;
        if (typeof this.nodeAux !== 'undefined') {
            bsLen += 2 * constants_1.ELEM_BYTES_LEN;
        }
        const arrBuff = new ArrayBuffer(bsLen);
        const bs = new Uint8Array(arrBuff);
        if (!this.existence) {
            bs[0] |= 1;
        }
        bs[1] = this.depth;
        bs.set(this.notEmpties, constants_1.PROOF_FLAG_LEN);
        const siblingBytes = (0, utils_1.siblings2Bytes)(this.siblings);
        bs.set(siblingBytes, this.notEmpties.length + constants_1.PROOF_FLAG_LEN);
        if (typeof this.nodeAux !== 'undefined') {
            bs[0] |= 2;
            bs.set(this.nodeAux.key.value, bs.length - 2 * constants_1.ELEM_BYTES_LEN);
            bs.set(this.nodeAux.value.value, bs.length - 1 * constants_1.ELEM_BYTES_LEN);
        }
        return bs;
    }
    allSiblings() {
        return (0, utils_1.siblignsFroomProof)(this);
    }
}
exports.Proof = Proof;
