// xitkov/lib/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotFound = exports.Forbidden = exports.PaymentRequired = exports.Unauthorized = exports.InternalServer = exports.NoContent = exports.BadRequest = void 0;
class BadRequest {
    constructor(message) {
        this.status = 400;
        throw { message, status: this.status };
    }
}
exports.BadRequest = BadRequest;
class NoContent {
    constructor(message) {
        this.status = 204;
        throw { message, status: this.status };
    }
}
exports.NoContent = NoContent;
class InternalServer {
    constructor(message) {
        this.status = 500;
        throw { message, status: this.status };
    }
}
exports.InternalServer = InternalServer;
class Unauthorized {
    constructor(message) {
        this.status = 401;
        throw { message, status: this.status };
    }
}
exports.Unauthorized = Unauthorized;
class PaymentRequired {
    constructor(message) {
        this.status = 402;
        throw { message, status: this.status };
    }
}
exports.PaymentRequired = PaymentRequired;
class Forbidden {
    constructor(message) {
        this.status = 403;
        throw { message, status: this.status };
    }
}
exports.Forbidden = Forbidden;
class NotFound {
    constructor(message) {
        this.status = 404;
        throw { message, status: this.status };
    }
}
exports.NotFound = NotFound;
